File: Classes/THView/THHighlightedTableView.m
// THHighlightedTableView.m

#import "THHighlightedTableView.h"
#import "THFoundation.h"
#import "THLog.h"
#import "TH_APP-Swift.h"

//------------------------------------------------------------------------------------------------------------------------------------------------------------------------
@implementation THHighlightedTableViewScrollView

- (void)scrollWheel:(NSEvent *)event
{
	[super scrollWheel:event];
//	[self updateHighLightedCellAfterScroll];
}

- (void)reflectScrolledClipView:(NSClipView *)aClipView
{
	[super reflectScrolledClipView:aClipView];
	[self updateHighLightedCellAfterScroll];
}

- (void)updateHighLightedCellAfterScroll
{
	if (_enclosedTableView==nil)
	{
		_enclosedTableView=(THHighlightedTableView*)self.documentView;
		THException([_enclosedTableView isKindOfClass:[THHighlightedTableView class]]==NO,@"documen view is not THHighlightedTableView class");
	}
	[_enclosedTableView updateHighLightedCellAfterScroll];

//	[(id<THHighlightedTableViewScrollViewDelegateProtocol>)self.delegatePageController highlightedTableViewScrollViewDidScroll:self];
}

@end
//------------------------------------------------------------------------------------------------------------------------------------------------------------------------


//------------------------------------------------------------------------------------------------------------------------------------------------------------------------
@implementation THHighlightedTableView

- (void)awakeFromNib
{
	[super awakeFromNib];

	THHighlightedTableViewScrollView *sv=(THHighlightedTableViewScrollView*)self.enclosingScrollView;
	THException([sv isKindOfClass:[THHighlightedTableViewScrollView class]]==NO,@"enclosingScrollView is not THHighlightedTableViewScrollView sv:%@",sv);
}

- (void)dealloc
{
	[self removeTrackingArea:_highlightedTrackingArea];
}

- (id<THHighlightedTableViewDelegateProtocol>)highlightedDelegate
{
	return (id<THHighlightedTableViewDelegateProtocol>)self.delegate;
}

#pragma mark -

- (void)reloadData
{
	_highlightedRow=-1;
	[super reloadData];
}

#pragma mark -

- (void)startHighlightedTracking
{
	if (_highlightedTrackingArea!=nil)
		return;
	_highlightedRow=-1;
	[self refreshHighlightedTracking];
}

- (void)stopHighlightedTracking
{
	_highlightedRow=-1;
	[self removeTrackingArea:_highlightedTrackingArea];
}

- (void)refreshHighlightedTracking
{
	[self removeTrackingArea:_highlightedTrackingArea];
	NSTrackingAreaOptions options=		NSTrackingMouseEnteredAndExited|NSTrackingMouseMoved|
																NSTrackingActiveAlways|
																NSTrackingInVisibleRect;
	_highlightedTrackingArea=[[NSTrackingArea alloc] initWithRect:NSMakeRect(0.0,0.0,self.frame.size.width,self.frame.size.height) options:options owner:self userInfo:nil];
	[self addTrackingArea:_highlightedTrackingArea];
}

- (void)updateTrackingAreas
{
	[super updateTrackingAreas];
	[self refreshHighlightedTracking];
}

#pragma mark -

- (void)updateHighLightedCellWithPoint:(NSPoint)point
{
	if (_firstHighlightedPoint.x==0.0 && _firstHighlightedPoint.y==0.0 && _highlightedRow==0)
		_highlightedRow=-1;

	if (_highlightedRow>=self.numberOfRows)
		_highlightedRow=-1;		

	_firstHighlightedPoint=point;
	NSInteger row=[self rowAtPoint:point];
	if (row==_highlightedRow)
		return;

	_highlightedRowIsChanging=YES;

//	if (self.selectedRow!=-1)
//		[self deselectRow:self.selectedRow];

	[[self highlightedDelegate] highlightedTableView:self didHighlightRow:row previousHighlightedRow:_highlightedRow];
	_highlightedRow=row;

	_highlightedRowIsChanging=NO;
}

#pragma mark -

- (BOOL)canHightlight
{
	if (self.window.isKeyWindow==YES || [[NSApplication sharedApplication] isActive]==NO)
		return YES;
	if ([[NSApplication sharedApplication] keyWindow]==self.window)
		return YES;
	return NO;
}

- (void)mouseEntered:(NSEvent*)event
{
	if ([[NSApplication sharedApplication] isActive]==NO)
		[[NSApplication sharedApplication] activateIgnoringOtherApps:YES];

	[self.window makeFirstResponder:self];

	if ([self canHightlight]==NO)
		return;
	
	NSPoint pt=[self convertPoint:[event locationInWindow] fromView:nil];
	[self updateHighLightedCellWithPoint:pt];
}

- (void)mouseExited:(NSEvent*)event
{
	if ([self canHightlight]==NO)
		return;

	[[self highlightedDelegate] highlightedTableView:self didHighlightRow:-1 previousHighlightedRow:_highlightedRow];
	_highlightedRow=-1;
}

- (void)mouseMoved:(NSEvent*)event
{
	if ([self canHightlight]==NO)
		return;

	NSPoint pt=[self convertPoint:[event locationInWindow] fromView:nil];
	[self updateHighLightedCellWithPoint:pt];
}

#pragma mark -

- (void)updateHighLightedRowFromSelectionDidChange
{
	if (_highlightedRowIsChanging==YES)
		return;

	NSInteger pRow=_highlightedRow;
	_highlightedRow=self.selectedRow;
	[[self highlightedDelegate] highlightedTableView:self didHighlightRow:_highlightedRow previousHighlightedRow:pRow];
}

- (void)updateHighLightedCellAfterScroll
{
	NSPoint location=[self.window mouseLocationOutsideOfEventStream];
	NSPoint pt=[self convertPoint:location fromView:nil];
	[self updateHighLightedCellWithPoint:pt];
}

- (void)updateHighLightedCellAfterDeletion
{
	_highlightedRow=-1;
	[self updateHighLightedCellAfterScroll];
}

- (void)unhighlightRow
{
	NSInteger pRow=_highlightedRow;
	_highlightedRow=-1;
	[[self highlightedDelegate] highlightedTableView:self didHighlightRow:_highlightedRow previousHighlightedRow:pRow];
}

#pragma mark -

- (NSPoint)convertWindowPointOfRow:(NSInteger)row
{
	NSRect rowRect=[self rectOfRow:row];

	NSPoint pt=NSMakePoint(rowRect.origin.x,rowRect.origin.y+CGFloatFloor(rowRect.size.height/2.0));
	pt=[self convertPointToBase:pt];
	//NSPoint p=[self.view.window.contentView convertPoint:rowPoint toView:self.tableView];
	pt=[self.window convertRectToScreen:NSMakeRect(pt.x,pt.y,0.0,0.0)].origin;

	return pt;
}

@end
//------------------------------------------------------------------------------------------------------------------------------------------------------------------------

File: THView/THHighlightedTableView.swift
// THHighlightedTableView.swift

import Cocoa

//------------------------------------------------------------------------------------------------------------------------------------------------------------------------
@objc class THHighlightedTableRowView : NSTableRowView {

	//@property (nonatomic,strong) NSColor *highlightedBackgroundColor;
	//@property (nonatomic,strong) NSColor *selectionBackgroundColor;

	@objc var isHighlightedRow = false { didSet {
		
		needsDisplay = true

		for view in self.subviews {
			if let cell = view as? THHighlightedTableCellView {
				cell.isHighlightedRow = isHighlightedRow
			}
		}

	}}

}
//------------------------------------------------------------------------------------------------------------------------------------------------------------------------


//------------------------------------------------------------------------------------------------------------------------------------------------------------------------
@objc class THHighlightedTableCellView : NSTableCellView {

	@objc var isHighlightedRow = false { didSet {
		updateHighlighted()
		}
	}

	func updateHighlighted() {
	}

}
//------------------------------------------------------------------------------------------------------------------------------------------------------------------------


//------------------------------------------------------------------------------------------------------------------------------------------------------------------------
@objc class THHighlightedTableViewScrollView: NSScrollView {

	private var mEnclosedTableView: THHighlightedTableView?
	
	override func scrollWheel(with event: NSEvent) {
		super.scrollWheel(with: event)
		//	[self updateHighLightedCellAfterScroll];
	}

	override func reflectScrolledClipView(_ cView: NSClipView) {
		super.reflectScrolledClipView(cView)
		updateHighLightedCellAfterScroll()
	}

	@objc func updateHighLightedCellAfterScroll() {
		if mEnclosedTableView == nil {
			mEnclosedTableView = self.documentView as? THHighlightedTableView
			THFatalError(mEnclosedTableView == nil, "mEnclosedTableView == nil")
		}
		mEnclosedTableView?.updateHighLightedCellAfterScroll()

		//	[(id<THHighlightedTableViewScrollViewDelegateProtocol>)self.delegatePageController highlightedTableViewScrollViewDidScroll:self];
	}

}
//------------------------------------------------------------------------------------------------------------------------------------------------------------------------


//------------------------------------------------------------------------------------------------------------------------------------------------------------------------
@objc protocol THHighlightedTableViewDelegateProtocol: NSTableViewDelegate {
	func highlightedTableView(_ tableView: THHighlightedTableView, didHighlightRow highlightedRow: Int, previousHighlightedRow: Int)
}

@objc class THHighlightedTableView: NSTableView {
	private var mHighlightedTrackingArea: NSTrackingArea?
	private var mFirstHighlightedPoint: NSPoint?
	private var mHighlightedRow = -1
	private var mHighlightedRowIsChanging = false

	private var highlightedDelegate: THHighlightedTableViewDelegateProtocol? { self.delegate as? THHighlightedTableViewDelegateProtocol }

	override func awakeFromNib() {
		super.awakeFromNib()

		THFatalError((self.enclosingScrollView is THHighlightedTableViewScrollView) == false, "unexpected kind of enclosingScrollView")
	}

	deinit {
		if let highlightedTrackingArea = mHighlightedTrackingArea {
			removeTrackingArea(highlightedTrackingArea)
		}
	}

	// MARK: -

	override func reloadData() {
		mHighlightedRow = -1
		super.reloadData()
	}

	// MARK: -

	@objc func startHighlightedTracking() {
		if mHighlightedTrackingArea != nil {
			return
		}

		mHighlightedRow = -1
		refreshHighlightedTracking()
	}

	@objc func stopHighlightedTracking() {
		mHighlightedRow = -1

		if let highlightedTrackingArea = mHighlightedTrackingArea {
			removeTrackingArea(highlightedTrackingArea)
		}
	}

	@objc func refreshHighlightedTracking() {
		if let highlightedTrackingArea = mHighlightedTrackingArea {
			removeTrackingArea(highlightedTrackingArea)
		}

		mHighlightedTrackingArea = NSTrackingArea(	rect: NSRect(0.0, 0.0, self.frame.size.width, self.frame.size.height),
																					options: [.mouseEnteredAndExited, .mouseMoved, .activeAlways, .inVisibleRect],
																					owner: self,
																					userInfo: nil)

		addTrackingArea(mHighlightedTrackingArea!)
	}

	override func updateTrackingAreas() {
		super.updateTrackingAreas()
		refreshHighlightedTracking()
	}

	// MARK: -

	@objc func updateHighLightedCell(withPoint point: NSPoint) {
		if mFirstHighlightedPoint == nil && mHighlightedRow == 0 {
			mHighlightedRow = -1
		}

		if mHighlightedRow >= self.numberOfRows {
			mHighlightedRow = -1
		}

		mFirstHighlightedPoint = point
		
		let row = self.row(at: point)
		if row == mHighlightedRow {
			return
		}

		mHighlightedRowIsChanging = true

	//	if (self.selectedRow!=-1)
	//		[self deselectRow:self.selectedRow];

		self.highlightedDelegate?.highlightedTableView(self, didHighlightRow: row, previousHighlightedRow: mHighlightedRow)
		mHighlightedRow = row
	
		mHighlightedRowIsChanging = false
	}

	// MARK: -

	@objc func canHightlight() -> Bool {
		if self.window?.isKeyWindow == true || NSApplication.shared.isActive == false {
			return true
		}
		if NSApplication.shared.keyWindow == self.window {
			return true
		}
		return false
	}

	override func mouseEntered(with event: NSEvent) {
		if NSApplication.shared.isActive == false {
			NSApplication.shared.activate(ignoringOtherApps: true)
		}

		self.window?.makeFirstResponder(nil)

		if self.canHightlight() == false {
			return
		}
		
		updateHighLightedCell(withPoint: self.convert(event.locationInWindow, from: nil))
	}
	
	override func mouseExited(with event: NSEvent) {
		if self.canHightlight() == false {
			return
		}

		self.highlightedDelegate?.highlightedTableView(self, didHighlightRow: -1, previousHighlightedRow: mHighlightedRow)
		mHighlightedRow = -1
	}

	override func mouseMoved(with event: NSEvent) {
		if self.canHightlight() == false {
			return
		}

		updateHighLightedCell(withPoint: self.convert(event.locationInWindow, from: nil))
	}

	// MARK: -

	@objc func updateHighLightedRowFromSelectionDidChange() {
		if mHighlightedRowIsChanging == true {
			return
		}

		let pRow = mHighlightedRow
		mHighlightedRow = self.selectedRow

		self.highlightedDelegate?.highlightedTableView(self, didHighlightRow: mHighlightedRow, previousHighlightedRow: pRow)
	}

	@objc func updateHighLightedCellAfterScroll() {
		guard let location = self.window?.mouseLocationOutsideOfEventStream
		else {
			return
		}
		updateHighLightedCell(withPoint: self.convert(location, from: nil))
	}

	@objc func updateHighLightedCellAfterDeletion() {
		mHighlightedRow = -1
		updateHighLightedCellAfterScroll()
	}

	@objc func unhighlightRow() {
		let pRow = mHighlightedRow
		mHighlightedRow = -1
		
		self.highlightedDelegate?.highlightedTableView(self, didHighlightRow: mHighlightedRow, previousHighlightedRow: pRow)
	}

	// MARK: -

	@objc func convertWindowPoint(forRow row: Int) -> NSPoint {
		let rowRect = self.rect(ofRow: row)

		var pt = NSPoint(rowRect.origin.x, rowRect.origin.y + (rowRect.size.height / 2.0).rounded(.down))
		pt = convertToBacking(pt)
		//NSPoint p=[self.view.window.contentView convertPoint:rowPoint toView:self.tableView];
		pt = self.window!.convertToScreen(NSMakeRect(pt.x,pt.y,0.0,0.0)).origin

		return pt
	}

}
//------------------------------------------------------------------------------------------------------------------------------------------------------------------------
